Only pass line property kwargs to a contour when plotting CompositeMap

ConorMacBride opened the pull request:
<!--
We know that working on code and submitting pull requests takes effort, and we appreciate your time.
Thank you.

Please be aware that everyone has to follow our code of conduct:
https://sunpy.org/coc

Furthermore, you might need to check with your work place if you are allowed to contribute code:
https://docs.sunpy.org/en/latest/dev_guide/contents/newcomers.html

Also these comments are hidden when you submit this github pull request.

We have several automatic features that run when a pull request is open.
They can appear daunting but do not worry about them!
Here is a brief explanation of them:
https://docs.sunpy.org/en/latest/dev_guide/contents/pr_review_procedure.html#continuous-integration
-->

### Description

This PR filters the matplotlib kwargs passed to a `CompositeMap.plot()` such that kwargs related to line style and line colour are only used when plotting contours. (Line width kwargs were already filtered.) Incorrect contour kwargs such as `linestyle`, `ls`, `color` and `c` are also filtered such that matplotlib raises a warning instead of an error.

<!--
Provide a general description of what your pull request does.

If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number GitHub will automatically link it.
If it doesn't, please remove the following line.
-->

Fixes #5455

pep8speaks commented:
Hello @ConorMacBride! Thanks for opening this PR. 



* In the file [`sunpy/map/tests/test_compositemap.py`](https://github.com/sunpy/sunpy/blob/dd888064f218e3e382d97c05c17ede7da114cf1f/sunpy/map/tests/test_compositemap.py):

> [Line 69:20](https://github.com/sunpy/sunpy/blob/dd888064f218e3e382d97c05c17ede7da114cf1f/sunpy/map/tests/test_compositemap.py#L69): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent

Do see the [Hitchhiker's guide to code style](https://goo.gl/hqbW4r)

dstansby commented:
There should be a cleverer way to do this, using the `inspect` module in the standard library. In particular, [`inspect.signature`](https://docs.python.org/3/library/inspect.html#introspecting-callables-with-the-signature-object) can be used to get the acceptable keyword arguments for a function or method. You should be able to use this to get the keyword arguments accepted by `GenericMap.plot` and `plt.imshow`, and then delete any keyword arguments not in that list.

I think we should also raise a warning that lists the keys being deleted.

ayshih commented:
> [`inspect.signature`](https://docs.python.org/3/library/inspect.html#introspecting-callables-with-the-signature-object) can be used to get the acceptable keyword arguments for a function or method.

You can't, because many of the keyword arguments aren't explicit in the signatures, but rather largely lumped in `**kwargs` (possibly to be passed on to lower-level functions).

dstansby commented:
Ah, I only went two steps down the `**kwargs` rabbit hole 😄 . It seems like it would be possible but for the chain ending at "`Artist` properties" here: https://matplotlib.org/stable/api/image_api.html#matplotlib.image.AxesImage, and I'm not sure how one could query for them.

ConorMacBride commented:
I think the following code works although it may be making too many assumptions about the inner workings of Matplotlib. It uses `inspect` to get valid keywords from the signature of various functions and methods. As you say, not all valid kwargs appear in a signature, however you can query the presence of callable `set_...` methods, which is how it's done internally (see https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/artist.py#L1058-L1061).

As you can see from the output of the code, we would have to do more filtering. For example `linestyles` and `linewidths` are valid for both `pcolormesh` and `contour`.

```py
import inspect

import matplotlib.pyplot as plt
from matplotlib.image import AxesImage, _ImageBase
from matplotlib.contour import ContourSet, QuadContourSet
from matplotlib.collections import QuadMesh, Collection


def get_keywords(func) -> set:
    """Returns a set of keyword names from `func`'s docstring.
    Recursive if `func` is a list of functions and methods."""
    if isinstance(func, list):
        keywords = set()
        for f in func:
            keywords.update(get_keywords(f))
        return keywords
    sig = inspect.signature(func)
    keywords = {param.name for param in sig.parameters.values()
                if param.default is not inspect.Parameter.empty}
    return keywords


def get_set_methods(obj) -> set:
    """Returns a set of keyword names that can be handled by 
    an object's `set_...` methods."""
    # See https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/artist.py#L1058-L1061
    return {
        m[4:] for m in dir(obj)
        if m.startswith('set_') and callable(getattr(obj, m, None))
    }


# Valid keyword arguments for each plotting method
imshow_kwargs = get_keywords(
    [plt.Axes.imshow, AxesImage.__init__, _ImageBase.__init__]
) | get_set_methods(AxesImage)

pcolormesh_kwargs = get_keywords(
    [plt.Axes.pcolormesh, QuadMesh.__init__, Collection.__init__]
) | get_set_methods(QuadMesh)

contour_kwargs = get_keywords(
    [ContourSet.__init__, QuadContourSet._process_args]
)

# Analysis

print("Valid image kwargs:\n\t", imshow_kwargs | pcolormesh_kwargs, '\n')

print("-----\n")
    
print("Valid image kwargs that...\n")
print("...would raise error for pcolormesh:\n\t", imshow_kwargs - pcolormesh_kwargs, '\n')
print("...would raise error for imshow:\n\t", pcolormesh_kwargs - imshow_kwargs, '\n')
print("...would raise error for contour:\n\t", (imshow_kwargs & pcolormesh_kwargs) - contour_kwargs, '\n')

print("-----\n")

print("Valid contour kwargs:\n\t", contour_kwargs, '\n')

print("-----\n")
    
print("Valid contour kwargs that...\n")
print("...would raise error for pcolormesh:\n\t", contour_kwargs - pcolormesh_kwargs, '\n')
print("...would raise error for imshow:\n\t", contour_kwargs - imshow_kwargs, '\n')

print("-----\n")

print("Valid kwargs that are valid for (imshow AND/OR pcolormesh) AND contour:")
print("\t", (imshow_kwargs | pcolormesh_kwargs) & contour_kwargs)
```

ConorMacBride commented:
Here's the output of the code above:
```
Valid image kwargs:
	 {'clip_box', 'zorder', 'aspect', 'resample', 'picker', 'clip_on', 'urls', 'interpolation', 'offsets', 'figure', 'ls', 'transOffset', 'gid', 'capstyle', 'clim', 'clip_path', 'visible', 'antialiaseds', 'url', 'label', 'norm', 'transform', 'filternorm', 'aa', 'shading', 'filterrad', 'cmap', 'agg_filter', 'linewidth', 'fc', 'vmax', 'data', 'antialiased', 'dashes', 'linewidths', 'vmin', 'linestyle', 'path_effects', 'array', 'alpha', 'facecolor', 'ec', 'origin', 'animated', 'snap', 'linestyles', 'offset_position', 'in_layout', 'pickradius', 'sketch_params', 'color', 'edgecolor', 'rasterized', 'facecolors', 'hatch', 'joinstyle', 'extent', 'contains', 'lw', 'paths', 'edgecolors'} 

-----

Valid image kwargs that...

...would raise error for pcolormesh:
	 {'filterrad', 'interpolation', 'extent', 'filternorm', 'aspect', 'resample', 'origin'} 

...would raise error for imshow:
	 {'linewidth', 'fc', 'dashes', 'antialiased', 'linewidths', 'urls', 'offsets', 'linestyle', 'facecolor', 'ls', 'ec', 'transOffset', 'capstyle', 'linestyles', 'offset_position', 'paths', 'pickradius', 'color', 'edgecolor', 'antialiaseds', 'facecolors', 'hatch', 'joinstyle', 'lw', 'aa', 'shading', 'edgecolors'} 

...would raise error for contour:
	 {'clip_box', 'zorder', 'agg_filter', 'data', 'picker', 'clip_on', 'figure', 'array', 'path_effects', 'gid', 'animated', 'snap', 'clim', 'clip_path', 'in_layout', 'visible', 'sketch_params', 'rasterized', 'url', 'label', 'contains'} 

-----

Valid contour kwargs:
	 {'cmap', 'levels', 'linestyles', 'filled', 'corner_mask', 'vmax', 'nchunk', 'antialiased', 'locator', 'extend', 'linewidths', 'hatches', 'vmin', 'colors', 'norm', 'extent', 'alpha', 'transform', 'origin'} 

-----

Valid contour kwargs that...

...would raise error for pcolormesh:
	 {'origin', 'hatches', 'levels', 'colors', 'filled', 'extent', 'corner_mask', 'extend', 'nchunk', 'locator'} 

...would raise error for imshow:
	 {'locator', 'linewidths', 'hatches', 'levels', 'linestyles', 'colors', 'filled', 'corner_mask', 'extend', 'nchunk', 'antialiased'} 

-----

Valid kwargs that are valid for (imshow AND/OR pcolormesh) AND contour:
	 {'origin', 'linewidths', 'cmap', 'vmin', 'linestyles', 'norm', 'extent', 'alpha', 'transform', 'vmax', 'antialiased'}
```

dstansby reviewed (changes requested):
For the sake of getting a fix in, how about we merge (and backport) this PR as is, and then we can think about taking the more complicated approach. Could you update the hashes for the devdeps figure test? You can find the correct ones in the build output here: https://app.circleci.com/pipelines/github/sunpy/sunpy/14270/workflows/d3b7dbf3-c086-4a16-baad-e14adbf7ec92/jobs/70694

ConorMacBride commented:
@dstansby Great! I've updated the figure hashes and added a changelog entry. When this is merged I can open a draft PR showing how it could be implemented.

nabobalis commented:
Thanks for the PR @ConorMacBride
